cro-clib/src/lib.rs

```diff
@@ -0,0 +1,263 @@
+use chain_core::init::network::init_chain_id;
+use client_common::error::ErrorKind;
+use client_common::storage::MemoryStorage;
+use client_common::SecKey;
+use client_core::types::WalletKind;
+use client_core::wallet::{DefaultWalletClient, WalletClient};
+use client_core::HDSeed;
+use client_core::Mnemonic;
+use secstr::SecUtf8;
+use std::ffi::CStr;
+use std::os::raw::c_char;
+use std::str::FromStr;
+pub const MAX_LENGTH: u32 = 1024;
+pub const MAX_STRING_LENGTH: u32 = 512;
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct Buf {
+    pub buf: [::std::os::raw::c_uchar; 512usize],
+    pub length: ::std::os::raw::c_int,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct ApiResult {
+    pub error: ::std::os::raw::c_int,
+    pub value: Buf,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct HDWallet {
+    pub error: ::std::os::raw::c_int,
+    pub name: Buf,
+    pub value: Buf,
+    pub mnemonics: Buf,
+    pub viewkey: Buf,
+    pub seed: Buf,
+    pub enckey: Buf,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct ApiContext {
+    pub error: ::std::os::raw::c_int,
+    pub chain_id: Buf,
+    pub server_url: Buf,
+    pub storage_folder: Buf,
+}
+
+static mut G_DB: Option<MemoryStorage> = None;
+
+/// # Safety
+/// this function should not be called inside rust
+unsafe fn get_string(src: *const c_char) -> String {
+    CStr::from_ptr(src).to_string_lossy().into_owned()
+}
+
+fn copy_string(src: &str, dst: &mut [u8]) {
+    dst[..src.len()].clone_from_slice(&src.as_bytes()[..src.len()]);
+    dst[src.len()] = 0;
+}
+
+fn get_string_from_buf(src: &Buf) -> String {
+    std::str::from_utf8(&src.buf[0..src.length as usize])
+        .expect("convert to string")
+        .to_string()
+}
+
+fn get_buf(src: &str) -> Buf {
+    let mut ret = Buf {
+        buf: [0; MAX_STRING_LENGTH as usize],
+        length: 0,
+    };
+    copy_string(src, &mut ret.buf);
+    ret.length = src.len() as i32;
+    ret
+}
+
+fn get_buf_vec(src: &[u8]) -> Buf {
+    let mut ret = Buf {
+        buf: [0; MAX_STRING_LENGTH as usize],
+        length: 0,
+    };
+    ret.buf[..src.len()].clone_from_slice(&src[..]);
+    ret.length = src.len() as i32;
+    ret
+}
+
+#[no_mangle]
+pub extern "C" fn get_network_id() -> ApiResult {
+    let networkid = chain_core::init::network::get_network_id();
+    let networkid_string = format!("{:2X}", networkid);
+    ApiResult {
+        error: 0,
+        value: get_buf(&networkid_string),
+    }
+}
+
+/// # Safety
+/// this function should not be called inside rust
+pub unsafe extern "C" fn do_create_hdwallet(
```

Comment:
For HD wallet, start with **stateless** functionality -- just like https://github.com/crypto-com/cro-nodelib/pull/14/files#diff-60a0e806fe678c1085ea774b2f44d37eR6

Just expose functions for generating seed / mnemonic and deriving addresses **without a global mutable state**

Reply:
ok, i got it.
global singleton will be removed.
user will maintain his own context.
chain_wallet_delete
chain_wallet_new

Reply:
wallet, account will opaque pointer.
which only exists as "struct .." , actually void*